cnudie/upload.py

```diff
@@ -0,0 +1,135 @@
+import enum
+import hashlib
+
+import ccc.oci
+import gci.componentmodel as cm
+import gci.oci
+import json
+import oci.model as om
+import dataclasses
+
+
+class UploadMode(enum.Enum):
+    SKIP = 'skip'
+    FAIL = 'fail'
+    OVERWRITE = 'overwrite'
+
+
+def _normalise_component_name(component_name: str) -> str:
+    return component_name.lower()  # oci-spec allows only lowercase
+
+
+def _target_oci_ref(
+    component: cm.Component,
+    component_ref: cm.ComponentReference=None,
+    component_version: str=None,
+):
+    if not component_ref:
+        component_ref = component
+        component_name = component_ref.name
+    else:
+        component_name = component_ref.componentName
+
+    component_name = _normalise_component_name(component_name)
+    component_version = component_ref.version
+
+    last_ctx_repo = component.current_repository_ctx()
+
+    return last_ctx_repo.component_version_oci_ref(
+        name=component_name,
+        version=component_version,
+    )
+
+
+def _sha_digest(fileobj) -> str:
+    digest = hashlib.sha256()
+    position = fileobj.tell()
+    while (chunk := fileobj.read(4096)):
+        digest.update(chunk)
+    fileobj.seek(position)
+    return f'sha256:{digest.hexdigest()}'
+
+
+def upload_component_descriptor(
+    component_descriptor: cm.ComponentDescriptor,
+    on_exist=UploadMode.SKIP,
+    ctx_repo_override: cm.OciRepositoryContext = None,
+):
+    schema_version = component_descriptor.meta.schemaVersion
+    if not schema_version is cm.SchemaVersion.V2:
+        raise RuntimeError(f'unsupported component-descriptor-version: {schema_version=}')
+
+    client = ccc.oci.oci_client()
+
+    if (
+        ctx_repo_override
+        and not component_descriptor.component.current_repository_ctx() == ctx_repo_override
+    ):
+        component_descriptor.component.repositoryContexts.append(ctx_repo_override)
+
+    target_ref = _target_oci_ref(component_descriptor.component)
+
+    if on_exist in (UploadMode.SKIP, UploadMode.FAIL):
+        # check whether manifest exists (head_manifest does not return None)
+        if client.head_manifest(image_reference=target_ref, absent_ok=True):
+            if on_exist is UploadMode.SKIP:
+                return
+            if on_exist is UploadMode.FAIL:
+                # XXX: we might still ignore it, if the to-be-uploaded CD is equal to the existing
+                # one
+                raise ValueError(f'{target_ref=} already existed')
+    elif on_exist is UploadMode.OVERWRITE:
+        pass
+    else:
+        raise NotImplementedError(on_exist)
+
+    raw_fobj = gci.oci.component_descriptor_to_tarfileobj(component_descriptor)
+    cd_sha_digest = _sha_digest(raw_fobj)
+    cd_octets = len(raw_fobj) # TODO: test
+
+    client.put_blob(
+        image_reference=target_ref,
+        digest=cd_sha_digest,
+        octets_count=cd_octets,
+        data=raw_fobj,
+        # mimetype=gci.oci.component_descriptor_mimetype, # not used in called func
```

Comment:
if it is not used, then remove it - or let us discuss changing the called function, if this seems appropriate. remembering the oci-transport-spec, I think blob-puts are always to use application/data, so it might really not make a lot of sense to specify a mimetype.